src/main/java/com/google/research/bleth/services/DatabaseService.java

```diff
@@ -0,0 +1,223 @@
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.repackaged.com.google.gson.Gson;
+import com.google.common.collect.ArrayTable;
+import com.google.common.collect.Multimap;
+import com.google.research.bleth.exceptions.BoardStateAlreadyExistsException;
+import com.google.research.bleth.exceptions.ExceedingRoundException;
+import com.google.research.bleth.simulator.AbstractSimulation;
+import com.google.research.bleth.simulator.Board;
+import com.google.research.bleth.simulator.IAgent;
+import com.google.research.bleth.simulator.Location;
+
+import java.util.ArrayList;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * DatabaseService is a singleton class wrapping the application's datastore instance.
+ * DatabaseService provides a global point of access to any read/write operation using the datastore instance.
+ *
+ * DatabaseService provides API for:
+ * - Writing simulation metadata by the builder responsible for the simulation construction (which returns the unique id
+ * assigned to the simulation, encoded to a string).
+ * - Writing a board state to the db.
+ * - Reading a board state from the db (which returns null if provided round exceeds simulation's max number of round,
+ * or a json string encoding the board state).
+ */
+public class DatabaseService {
+
+    private static DatabaseService instance = null;
+    private DatastoreService datastore;
+    private Gson gson = new Gson();
+
+    /**
+     * Static method to create an instance of the DatabaseService class.
+     * @return a DatabaseService class instance.
+     */
+    public static DatabaseService getInstance() {
+        if (instance == null) {
+            instance = new DatabaseService();
+        }
+        return instance;
+    }
+
+    /**
+     * Write an entity providing simulation metadata.
+     * @param builder is the builder responsible for the simulation construction.
+     * @return the new simulation id (encoded to string).
+     */
+    public String writeMetadata(AbstractSimulation.Builder builder) {
+        // Create new entity.
+        Entity entity = new Entity("Simulation");
+
+        // Set properties.
+        entity.setProperty("type", builder.getSimulationType());
+        entity.setProperty("roundsNum", builder.getMaxNumberOfRounds());
+        entity.setProperty("beaconsNum", builder.getBeaconsNum());
+        entity.setProperty("observersNum", builder.getObserversNum());
+        entity.setProperty("rowNum", builder.getRowNum());
+        entity.setProperty("colNum", builder.getColNum());
+        entity.setProperty("beaconMovementStrategy", builder.getBeaconMovementStrategy().toString());
+        entity.setProperty("observerMovementStrategy", builder.getObserverMovementStrategy().toString());
+        entity.setProperty("observerAwakenessStrategy", builder.getAwakenessStrategyType().toString());
```

Comment:
Do all the observers have the same strategy? If so - it's observer**s** and not observer, otherwise we should save this information about each agent.